### PR TITLE
Product sorting 

### PR DESCRIPTION
**Pull request by manfield**
Fixed product sorting (according to position) when processed by ps_facetedsearch module as Product Search provider.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/ps_facetedsearch/24)

### CODE REVIEW

**Comment by tomlev** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[ps_facetedsearch.php, line 1548 at r1](https://reviewable.io:443/reviews/prestashop/ps_facetedsearch/24#-KwARDmrqfR0dE5r-e-e:-KwARDmrqfR0dE5r-e-f:b-52u44q) ([raw file](https://github.com/prestashop/ps_facetedsearch/blob/03aa3b85f31d434ff3d325e406d05742a2b16dfb/ps_facetedsearch.php#L1548)):*
> ```PHP
>                     ' LIMIT '.(((int) $page - 1) * $products_per_page.','.$products_per_page), true, false);
>                     
>                     
> ```

useless spaces (not blocker)

---

*[ps_facetedsearch.php, line 1589 at r1](https://reviewable.io:443/reviews/prestashop/ps_facetedsearch/24#-KwARHlG5ABWJZVjR8i_:-KwARHlG5ABWJZVjR8ia:b-52u44q) ([raw file](https://github.com/prestashop/ps_facetedsearch/blob/03aa3b85f31d434ff3d325e406d05742a2b16dfb/ps_facetedsearch.php#L1589)):*
> ```PHP
>             'count' => $this->nbr_products,
>         );
>       
> ```

useless spaces (not blocker)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/ps_facetedsearch/24)*

**Comment by tomlev** [comment]
thank you @manfield

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/ps_facetedsearch/24#-:-KwARwVswiBK0sxOnoiZ:b-8ku5bw)*

**Comment by jocel1** [comment]
I'll check if I can integrate the changes directly into https://github.com/PrestaShop/ps_facetedsearch/pull/25 since it changes the same lines.

File: ps_facetedsearch.php [hunk 1]
```diff
@@ -1438,27 +1438,27 @@ public function getProductByFilters(
...
+                                                        STRAIGHT_JOIN '._DB_PREFIX_.'category c ON (c.id_category = cp.id_category AND
                                                         '.($this->ps_layered_full_tree ? 'c.nleft >= '.(int) $parent->nleft.'
                                                         AND c.nright <= '.(int) $parent->nright : 'c.id_category = '.(int) $id_parent).'
                                                         AND c.active = 1)
                                                         STRAIGHT_JOIN `'._DB_PREFIX_.'product` p ON (p.id_product=cp.id_product)
                                                         '.$price_filter_query_in.'
                                                         '.$query_filters_from.'
                                                         WHERE 1 '.$query_filters_where.'
-                                                        GROUP BY cp.id_product ORDER BY position, id_product', false);
+                                                        GROUP BY cp.id_product ORDER BY cp.position, cp.id_product', false);
```
Comment (jocel1) [inline ps_facetedsearch.php]: with the GROUP BY, if you do not use MIN(cp.position) you don't know which position you will get if there are several time the same id_product in different cat, but with different position. The current code makes sure you always use the smallest position from all possible cats, but with your code, your using a random position.

File: ps_facetedsearch.php [hunk 2]
```diff
@@ -1484,14 +1484,14 @@ public function getProductByFilters(
                         AND c.active = 1)
                     '.$price_filter_query_out.'
                     '.$query_filters_from.'
-                    WHERE 1 '.$query_filters_where.' GROUP BY cp.id_product');
+                    WHERE 1 '.$query_filters_where.' GROUP BY cp.id_product ORDER BY cp.position ASC');
```
Comment (jocel1) [inline ps_facetedsearch.php]: ORDER BY is useless, we are return a list of id, we don't care to have it ordered or not

File: ps_facetedsearch.php [hunk 3]
```diff
@@ -1484,14 +1484,14 @@ public function getProductByFilters(
...
             } else {
                 $all_products_out = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS('
                     SELECT p.`id_product` id_product
                     FROM `'._DB_PREFIX_.'product` p JOIN '._DB_PREFIX_.'category_product cp USING (id_product)
                     '.$price_filter_query_out.'
                     '.$query_filters_from.'
-                    WHERE cp.`id_category` IN ('.implode(',', $categories).') '.$query_filters_where.' GROUP BY cp.id_product');
+                    WHERE cp.`id_category` IN ('.implode(',', $categories).') '.$query_filters_where.' GROUP BY cp.id_product ORDER BY cp.position ASC');
```
Comment (jocel1) [inline ps_facetedsearch.php]: same here

**Comment by manfield** [comment]
Hi,
the main issue was line 1443:

`STRAIGHT_JOIN '._DB_PREFIX_.'category_product cp ON `

Let me check other points and I will be back to you

**Comment by jocel1** [comment]
@manfield :`'._DB_PREFIX_.'category_product cp STRAIGHT_JOIN '._DB_PREFIX_.'category c ` AND `'._DB_PREFIX_.'category c STRAIGHT_JOIN '._DB_PREFIX_.'category_product cp ` should return the same return as it's a simple join, it only changes potentially the join speed.

**Comment by manfield** [comment]
I had an issue on that query. I will come back with more details.
It has been a while I did not check that.

**Comment by jocel1** [comment]
But you could check if you still have the issue with the code in #25?

**Comment by manfield** [comment]
I do now

**Comment by manfield** [comment]
Actually it does not work.
This is the product sorting in backend:



This is with my code:


This is with #25 code

**Comment by jocel1** [comment]
what's the sorting inside the Women category ? (since it's the category you're requesting in the facetedsearch)

**Comment by manfield** [comment]
Hi,
it is exactly the one I posted in the screenshot.

**Comment by jocel1** [comment]
thanks a lot I've reproduced the issue locally, I'm taking a look at the code right now!

**Comment by jocel1** [comment]
FYI your fix is not correctly, but the issue is when the ps_layered_full_tree mode is activate, it also checked the product order in the subcategory, and if it's lower than the current cat (I assume in your case in the TOPS & DRESSES categories, there's no position set, so the value is 0), it override the value of the current cat. We need to take into account into the query that the position of the current cat should be used in priority if the product is also present in subcategories

**Comment by jocel1** [comment]
I've just pushed a fix, let me know if it works better for you

**Comment by eternoendless** [comment]
Closing this one as it's fixed in https://github.com/PrestaShop/ps_facetedsearch/pull/25

**Pull request closed without merging**